feat: add ability in robyn cli to scaffold example programs with various DBs

bwdq opened the pull request:
Fixes #511 and adds some example code for scaffolding Sqlite, Postgres, Mongo, Prisma, and SQLAlchemy.

sansyrox commented:
Hey @bwdq 👋 

Great work 😄 

I will have a thorough look over the weekend but this great work so far!

sansyrox commented on robyn/scaffold/postgres.py:
```diff
@@ -0,0 +1,40 @@
+if __name__ == "__main__":
+    app.start()
+
+# docker run
+#     --name some-postgres
+#     -p 5455:5432
+#     -e POSTGRES_USER=postgres
+#     -e POSTGRES_PASSWORD=password
+#     -e POSTGRES_DB=postgresDB
+#     -d
+#     postgres
```
Do we need these?

sansyrox reviewed (commented):
Hey @bwdq 👋 

Great work 😄 

Can you please bring the dockerfile in the scaffold folder instead of generating it on the fly?

sansyrox commented on robyn/scaffold/mongo.py:
All these databases would also require certain dependencies. Can we update the docker files accordingly?

bwdq replied:
I have updated the Docker files for the databases. When I ran docker it errored 'uvloop not found' so uvloop should be a mandatory dependency instead of optional.

sansyrox commented on robyn/__main__.py:
```diff
@@ -17,10 +17,24 @@ def create_robyn_app():
+        },
     ]
     result = prompt(questions=questions)
     project_dir = result[0]
     docker = result[1]
+    project_type = result[2]
```
```suggestion
    project_dir, docker, project_type = result
```

bwdq replied:
I tried this and it did not parse correctly so I left it as is

sansyrox replied:
@bwdq , what do you mean by this? What error were you getting?

We can return a dictionary instead of a tuple if there is an issue

sansyrox commented on robyn/__main__.py:
```diff
@@ -43,9 +72,8 @@ def index():
-
             """
-        )
+            )
```
do we need this?

bwdq replied:
I think my IDE did that. Outdated by my changing Dockerfile generation.

sansyrox commented on robyn/scaffold/docker/Dockerfile-mongo:
Hey @bwdq 👋 

Can we please move them to something like `/scaffold/mongo/Dockerfile` and the code to `/scaffold/mongo/app.py`

bwdq replied:
done

sansyrox commented on robyn/scaffold/prisma/app.py:
```diff
@@ -0,0 +1,35 @@
+
+# install prisma cli
+# initialize prisma with "prisma generate"
+# then "prisma migrate dev" and name migration dev
```
@bwdq 👋 

Can you please explain the line here?

bwdq replied:
These are instructions to the user because prisma won't "just work" and requires some setup. I'm not sure what the optimal solution here is. I don't want to execute bash commands silently.

sansyrox replied:
hey @bwdq 👋 

We can add a small readme inside every folder. Doesn't need to be too fancy. Can just contain these commands.

bwdq replied:
I have moved the python comments to README's.

sansyrox commented on robyn/scaffold/sqlite/app.py:
```diff
@@ -0,0 +1,21 @@
+    res = cur.execute("SELECT name FROM sqlite_master")
+    th = res.fetchone()
+    print(th)
+    return "Hello World!"
```
I believe that we should be including `res` in the final response.

bwdq replied:
done

sansyrox commented on robyn/__main__.py:
```diff
@@ -30,44 +44,48 @@ def create_robyn_app():
-
-            """
-        )
+        if project_type == "sqlite":
+            f.write(open("robyn/scaffold/sqlite/app.py", "r").read())
+        elif project_type == "postgres":
+            f.write(open("robyn/scaffold/postgres/app.py", "r").read())
+            # copy README.md
+            readme_path = os.path.join(project_dir, "README.md")
+            with open(readme_path, "w") as f_r:
+                f_r.write(open("robyn/scaffold/postgres/README.md", "r").read())
+        elif project_type == "mongo":
+            f.write(open("robyn/scaffold/mongo/app.py", "r").read())
+        elif project_type == "sqlalchemy":
+            f.write(open("robyn/scaffold/sqlalchemy/app.py", "r").read())
+        elif project_type == "prisma":
+            f.write(open("robyn/scaffold/prisma/app.py", "r").read())
+            # copy schema.prisma
+            schema_path = os.path.join(project_dir, "schema.prisma")
+            with open(schema_path, "w") as f_s:
+                f_s.write(open("robyn/scaffold/prisma/schema.prisma", "r").read())
+            # copy README.md
+            readme_path = os.path.join(project_dir, "README.md")
+            with open(readme_path, "w") as f_r:
+                f_r.write(open("robyn/scaffold/prisma/README.md", "r").read())
+        else:
+            f.write(open("robyn/scaffold/no-db/app.py", "r").read())
```
I wonder if there is a cleaner way to implement this?

can you store the file_name and file_paths in a dictionary and then try accessing it?

@IdoKendo do you have any thoughts on this?

IdoKendo replied:
I would simplify this whole block to something like these 3 lines:
```py
base_path = f"robyn/scaffold/{project_type}"
shutil.copyfile(f"{base_path}/README.md", readme_path)
shutil.copyfile(f"{base_path}/app.py",  app_file_path)
```
Just make sure the `project_type` default is `no-db` 😃 
Can add some guard clauses to be extra safe, but I think `InquirerPy` does that for you out-of-the-box.

IdoKendo replied:
Upon further review of the code, can even do it as simple as:
```py
from distutils.dir_util import copy_tree

copy_tree(project_dir, f"robyn/scaffold/{project_type}")
```

bwdq replied:
This is much simpler but at the time I didn't want to add a new dependency. I've switched to this now.

sansyrox commented on robyn/__main__.py:
```diff
@@ -30,44 +44,48 @@ def create_robyn_app():
-CMD ["python3.10", "/workspace/foo/app.py", "--log-level=DEBUG"]
-                """
-            )
+            if project_type == "sqlite":
+                f.write(open("robyn/scaffold/sqlite/Dockerfile", "r").read())
+            elif project_type == "postgres":
+                f.write(open("robyn/scaffold/postgres/Dockerfile", "r").read())
+            elif project_type == "mongo":
+                f.write(open("robyn/scaffold/mongo/Dockerfile", "r").read())
+            elif project_type == "sqlalchemy":
+                f.write(open("robyn/scaffold/sqlalchemy/Dockerfile", "r").read())
+            elif project_type == "prisma":
+                f.write(open("robyn/scaffold/prisma/Dockerfile", "r").read())
+            else:
```
same for this

IdoKendo reviewed (changes requested):
Hey @bwdq 👋 Nice job!
I've added some comments to simplify and improve upon the changes. LMK if there's anything I can help with!

sansyrox commented:
Thanks for the review @IdoKendo 😄 

I wasn't aware about this feature 

> from distutils.dir_util import copy_tree

sansyrox commented:
Any updates on this @bwdq ?

bwdq commented on robyn/scaffold/mongo/Dockerfile:
```diff
@@ -0,0 +1,14 @@
+FROM ubuntu:22.04
```
I have updated the base docker image and it would be helpful if you could test it.

bwdq commented on robyn/__main__.py:
```diff
@@ -17,10 +17,24 @@ def create_robyn_app():
+                Choice("sqlalchemy", name="SqlAlchemy"),
+                Choice("prisma", name="Prisma"),
+            ],
+            "default": None,
```
done

bwdq commented:
I have updated and simplified the logic in main. I'm still having issues with the robyn resetting the connection when running on docker and some testing would be helpful.

sansyrox commented:
Hey @bwdq 👋 

> I'm still having issues with the robyn resetting the connection when running on docker and some testing would be helpful.

Can you please explain what you mean here? I don't understand what you mean by "resetting" the connection.

bwdq commented:
The issue isn't really related to this PR because it's reproducible on the main branch so I've opened a separate issue for it.
https://github.com/sparckles/robyn/issues/621

sansyrox commented:
Hey @bwdq 👋 

Is the PR up for review?

bwdq commented:
> Hey @bwdq 👋
> 
> Is the PR up for review?

Yes it is. The docker issue is now fixed.

sansyrox commented:
Hey @bwdq 👋 

I have two suggestions for this PR
- there are some conflicts. Can we please resolve them?
- can we add some documentation regarding the new change in the `new_docs` folder?

bwdq commented on robyn/scaffold/no-db/Dockerfile:
```diff
@@ -0,0 +1,11 @@
+FROM python:3.11-bookworm
```
python dockerfile because that way I didn't have to install pip
3.11 was the latest stable version
bookworm is the latest stable debian. I tried Alpine and it didn't work

sansyrox reviewed (commented):
Hey @bwdq 👋 

I was going trying your branch and when I run the command `python -m robyn --create`

I get the following error

```
Please select project type (Mongo/Postgres/Sqlalchemy/Prisma):  Sqlite
Creating a new Robyn project '.'...
Traceback (most recent call last):
  File "/Users/sanskar/.pyenv/versions/3.10.7/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/Users/sanskar/.pyenv/versions/3.10.7/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/Users/sanskar/repos/robyn/robyn/__main__.py", line 68, in <module>
    create_robyn_app()
  File "/Users/sanskar/repos/robyn/robyn/__main__.py", line 51, in create_robyn_app
    copy_tree(f"robyn/scaffold/{project_type}/", project_dir)
  File "/Users/sanskar/repos/robyn/venv/lib/python3.10/site-packages/setuptools/_distutils/dir_util.py", line 135, in copy_tree
    raise DistutilsFileError("cannot copy tree '%s': not a directory" % src)
distutils.errors.DistutilsFileError: cannot copy tree 'robyn/scaffold/sqlite/': not a directory
```

And this is happening for every selection. Can you please have a look?

sansyrox commented:
Hey @bwdq 👋 

This PR has been open for a while now. I will be building upon this to get it landed!

sansyrox commented:
Hey @IdoKendo @Mr-Sunglasses 👋 

I hope all is well 😄 Would it be possible for you folks to have a final review at this PR?

IdoKendo commented on robyn/scaffold/sqlalchemy/app.py:
```diff
@@ -0,0 +1,32 @@
+
+if __name__ == "__main__":
+    # create a configured "Session" class
+    Base = declarative_base()
```
Maybe it's better to move the sqlalchemy stuff to another module? models.py perhaps?

IdoKendo reviewed (changes requested):
Hey @sansyrox 👋 I think it looks good! I would just make one last change with the SQLAlchemy template and I think we can merge this. 🚀

Mr-Sunglasses commented on robyn/scaffold/postgres/requirements.txt:
```diff
@@ -0,0 +1,2 @@
+robyn
+psycopg2
```
As `pip3 install psycopg2` did not work due to this [reason](https://stackoverflow.com/questions/5420789/how-to-install-psycopg2-with-pip-on-python), so we should change it to `psycopg2-binary` from `psycopg2`


Error when you only run `pip3 install psycopg2`
```
error: subprocess-exited-with-error

  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [23 lines of output]
      running egg_info
      creating /private/var/folders/p8/gp_5g0jx34vc15g471jmpnpw0000gn/T/pip-pip-egg-info-qzetowdt/psycopg2.egg-info
      writing /private/var/folders/p8/gp_5g0jx34vc15g471jmpnpw0000gn/T/pip-pip-egg-info-qzetowdt/psycopg2.egg-info/PKG-INFO
      writing dependency_links to /private/var/folders/p8/gp_5g0jx34vc15g471jmpnpw0000gn/T/pip-pip-egg-info-qzetowdt/psycopg2.egg-info/dependency_links.txt
      writing top-level names to /private/var/folders/p8/gp_5g0jx34vc15g471jmpnpw0000gn/T/pip-pip-egg-info-qzetowdt/psycopg2.egg-info/top_level.txt
      writing manifest file '/private/var/folders/p8/gp_5g0jx34vc15g471jmpnpw0000gn/T/pip-pip-egg-info-qzetowdt/psycopg2.egg-info/SOURCES.txt'

      Error: pg_config executable not found.

      pg_config is required to build psycopg2 from source.  Please add the directory
      containing pg_config to the $PATH or specify the full executable path with the
      option:

          python setup.py build_ext --pg-config /path/to/pg_config build ...

      or with the pg_config option in 'setup.cfg'.

      If you prefer to avoid building psycopg2 from source, please install the PyPI
      'psycopg2-binary' package instead.

      For further information please check the 'doc/src/install.rst' file (also at
      <https://www.psycopg.org/docs/install.html>).

      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.

```

sansyrox commented on robyn/__main__.py:
```diff
@@ -4,82 +4,71 @@
 from InquirerPy.base.control import Choice
 from .argument_parser import Config
 from robyn.robyn import get_version
+from distutils.dir_util import copy_tree
```
To fix
```
/home/batman/repos/robyn/robyn/__main__.py:7: DeprecationWarning: The distutils package is deprecated and slated for removal in Python 3.12. Use setuptools or check PEP 632 for potential alternatives
  from distutils.dir_util import copy_tree
```

Mr-Sunglasses reviewed (commented):
My two cents is that if we are adding a sqlalchemy template, then it will be a good idea to add a template of sqlalchemy with migrations ( alembic migrations )

WDYT @sansyrox @IdoKendo

sansyrox commented:
@Mr-Sunglasses , can you please explain a bit more? What would be different?

IdoKendo commented:
> My two cents is that if we are adding a sqlalchemy template, then it will be a good idea to add a template of sqlalchemy with migrations ( alembic migrations )
> 
> 
> 
> WDYT @sansyrox @IdoKendo 

Not sure we should hold this PR for this. after it's merged I think we can open a separate issue for this.

Mr-Sunglasses commented:
> > My two cents is that if we are adding a sqlalchemy template, then it will be a good idea to add a template of sqlalchemy with migrations ( alembic migrations )
> > WDYT @sansyrox @IdoKendo
> 
> Not sure we should hold this PR for this. after it's merged I think we can open a separate issue for this.

I agree with this, We should implement it when we migrate to a new repo #659 .

Mr-Sunglasses commented:
> @Mr-Sunglasses , can you please explain a bit more? What would be different?

@sansyrox I'm taking about sqlalchemy + alembic migration template but we can do after we migrate create_robyn_app to another repo ( #659 )

Mr-Sunglasses commented on robyn/__main__.py:
```diff
@@ -4,82 +4,75 @@
 
-EXPOSE 8080
+    selected_project_template = (SCAFFOLD_DIR / Path(project_type)).resolve()
+<<<<<<< HEAD
```
I think it is here by mistake, it should not be here 😢 . It'll cause errors.

sansyrox replied:
done!

sansyrox commented:
> > @Mr-Sunglasses , can you please explain a bit more? What would be different?
> 
> @sansyrox I'm taking about sqlalchemy + alembic migration template but we can do after we migrate create_robyn_app to another repo ( #659 )

@Mr-Sunglasses , can you please make an issue for the same?

Mr-Sunglasses commented:
> > > @Mr-Sunglasses , can you please explain a bit more? What would be different?
> > 
> > 
> > @sansyrox I'm taking about sqlalchemy + alembic migration template but we can do after we migrate create_robyn_app to another repo ( #659 )
> 
> @Mr-Sunglasses , can you please make an issue for the same?

Sure.

Mr-Sunglasses reviewed (approved):
LGTM Great work Team 🙌🏻

IdoKendo reviewed (approved):
LGTM 🔥

sansyrox commented:
Thank you @IdoKendo and @Mr-Sunglasses 😄 ✨
